packages/schematics/angular/utility/workspace.ts

```diff
@@ -10,60 +10,109 @@ import { json, workspaces } from '@angular-devkit/core';
 import { Rule, Tree, noop } from '@angular-devkit/schematics';
 import { ProjectType } from './workspace-models';
 
-function createHost(tree: Tree): workspaces.WorkspaceHost {
-  return {
-    async readFile(path: string): Promise<string> {
-      return tree.readText(path);
-    },
-    async writeFile(path: string, data: string): Promise<void> {
-      return tree.overwrite(path, data);
-    },
-    async isDirectory(path: string): Promise<boolean> {
-      // approximate a directory check
-      return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
-    },
-    async isFile(path: string): Promise<boolean> {
-      return tree.exists(path);
-    },
-  };
+const DEFAULT_WORKSPACE_PATH = '/angular.json';
+
+// re-export the workspace definition types for convenience
+export type WorkspaceDefinition = workspaces.WorkspaceDefinition;
+export type ProjectDefinition = workspaces.ProjectDefinition;
+export type TargetDefinition = workspaces.TargetDefinition;
+
+/**
+ * A {@link workspaces.WorkspaceHost} backed by a Schematics {@link Tree} instance.
+ */
+class TreeWorkspaceHost implements workspaces.WorkspaceHost {
+  constructor(private readonly tree: Tree) {}
+
+  async readFile(path: string): Promise<string> {
+    return this.tree.readText(path);
+  }
+
+  async writeFile(path: string, data: string): Promise<void> {
+    if (this.tree.exists(path)) {
+      this.tree.overwrite(path, data);
+    } else {
+      this.tree.create(path, data);
+    }
+  }
+
+  async isDirectory(path: string): Promise<boolean> {
+    // approximate a directory check
+    return !this.tree.exists(path) && this.tree.getDir(path).subfiles.length > 0;
+  }
+
+  async isFile(path: string): Promise<boolean> {
+    return this.tree.exists(path);
+  }
```

Comment:
Similarly here: `isFile('some/non/existant/path')` returns `false`, when I think most users would expect a non-existent file to throw or something. Not sure if this is fixable right now though.

Reply:
`isFile`/`isDirectory` are not expected to throw.  If the path does not reference the expected item, it should return `false`.  The `fs` based implementation also behaves in this manner.  We can review and/or adjust the workspace related APIs moving forward though.

Reply:
+1